app/views/my/solutions/solve.html.haml

```diff
@@ -0,0 +1,105 @@
+.lo-container
+  #solve-interface
+    #ace-editor
+      :preserve
+        class TwoFer
+          def self.two_fer(name="you")
+            "One for %s, one for me." % name
+          end
+        end
+
+    #tests-output
+      .ops-status
+        Ops Status:
+        %span
+      .results-status
+        Tests Status:
+        %span
+      .results-message
+        Message:
+        %span
+      .test-failure
+        Failed test:
+        %span.name
+        .message
+
+  %button#run-tests-button.pure-button Run Tests
+
+:scss
+  #solve-interface {
+    display:flex;
+  }
+
+  #ace-editor {
+    width:50%;
+    height:300px;
+    margin-top:10px;
+    margin-bottom:10px;
+  }
+  #tests-output {
+    padding-left:50px;
+
+    .ops-status,
+    .results-status,
+    .results-message,
+    .test-failure {
+      display:none;
+    }
+  }
+  #run-tests-button {
+    background:green;
+  }
+
+-content_for :js do
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/ace/1.4.6/ace.js" type="text/javascript" charset="utf-8"></script>
+  :javascript
+    // Set interval tracker for use later
+    // Use an array to avoid races
+    var intervals = {}
+
+    var editor = ace.edit("ace-editor");
+    editor.setTheme("ace/theme/monokai");
+    editor.session.setMode("ace/mode/#{@track.slug}");
+
+    $('#run-tests-button').click(function() {
+      var code = editor.getValue();
+      $.ajax(
+        "/my/solutions/#{@solution.uuid}/submit",
+        {
+          method: "PATCH",
+          data: {
+            files: {
+              "two_fer.rb": code
+            }
+          }
+        }
+      ).done(function(data) {
+        submissionId = data.submission.id
+        $('#tests-output .ops-status').show().children('span').html("Tests pending")
+
+        for(sid in intervals) { clearInterval(intervals[sid]) }
+
+        intervals[submissionId] = setInterval(function() {
```

Comment:
This does NOT what you want it to do. I'll explain:

If I understand correctly you want to:
- submit the code
- show pending
- each 500 ms check if it's done by getting the test results
- if it is, stop the interval and be done

But the way this is written is not optimal. It's:
- relying on the the request to _fail_ when it's not done
- has various race-conditions.

Consider the following. Let's say it takes 1.1s for test_results to return. It takes 2 seconds for the tests to run

- t: 0 .done code is executed
- t: 500: ➡ first ajax request to get results is _started_ 
- t: 1000: ➡ second ajax request to get results is _started_
- t: 1500: ➡ third ajax request to get results is _started_
- t: 1600: 🔙 first ajax request is _received_, this is an error, ❌ throw away
- t: 2000: ➡ fourth ajax request to get results is _started_
- t: 2100: 🔙 second ajax request is _received_, this is an error, ❌ throw away
- t: 2500: ➡ fifth ajax request to get results is _started
- t: 2600: 🔙 third ajax request  is _received_, this MIGHT have the result. Let's say it does. ✔ executes callback, no new ajax requests will be scheduled
- t: 3100: 🔙 fourth ajax request  is _received_, with result ✔ executes callback
- t: 3600: 🔙 fifth ajax request  is _received_, with result ✔ executes callback

A quick fix, to stop the "double renders/executes" is to, at line 87, break out of the function:

```javascript
              if(intervals[submissionId]) { 
                clearInterval(intervals[submissionId])
                delete intervals[submissionId]
              } else {
                return /* double render */
              }
```

... but better would be to use `setTimeout` instead, and `onError`, start a new request. Yes, it will take _up to a roundtrip_ longer to see the results, but it will save you so many requests:

- t: 0 .done code is executed
- t: 500: ➡ first ajax request to get results is _started_ 
- t: 1600: 🔙 first ajax request is _received_, this is an error, ❌ throw away
- t: 1600: ➡ error handler for ajax starts new request 
- t: 2700: 🔙 second ajax request is _received_, this MIGHT have the result. Let's say it does. ✔ executes callback, no new ajax requests will be scheduled

If calls come back _within 500 ms, you can also choose to schedule the next one _a but further out_:

```javascript
whenError((response) => {
  const timeSinceLastRequest = convertToUnix(response['Date']) - previousRequestTime
  intervals[submissionId] = setTimeout(requestResults, Math.max(500 - timeSinceLastRequest, 0))
})
```